modules/transaction-payment/src/lib.rs

```diff
@@ -611,59 +673,206 @@ where
+		Self::swap_native_asset(who, amount);
+	}
+
+	/// Iterate order list, break if can swap out enough native asset amount with user's foreign
+	/// asset. make sure trading path is exist in dex, if the trading pair is not exist in dex, even
+	/// though we have setup it in charge fee pool, we can't charge fee with this foreign asset.
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			if let Some(target_currency_id) = trading_path.last() {
+				if *target_currency_id == native_currency_id {
+					let supply_currency_id = *trading_path.first().expect("should match a non native asset");
+					if Self::swap_from_treasury_or_dex(who, amount, supply_currency_id).is_ok() {
+						break;
+					}
+				}
+			}
+		}
+	}
+
+	/// swap treasury account's native asset with user's foreign asset from treasury fee pool when
+	/// treasury account's native asset balance is gt threshold. or-else, swap treasury account's
+	/// native asset with treasury account's foreign asset from dex swap liquid pool.
+	#[transactional]
+	fn swap_from_treasury_or_dex(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> sp_std::result::Result<(), DispatchError> {
+		match TokenFixedRate::<T>::get(supply_currency_id) {
+			None => {
+				// if the token rate is none(i.e. disable by council or on purpose), return err.
+				Err(DispatchError::CannotLookup)
+			}
+			Some(mut rate) => {
+				// user's supply asset should have enough balance based on old fix rate.
+				ensure!(
+					T::MultiCurrency::free_balance(supply_currency_id, who) >= rate.saturating_mul_int(amount),
+					Error::<T>::SupplyBalanceTooLow
+				);
+				let treasury_account = Self::treasury_sub_account_id(supply_currency_id);
+
+				// if treasury account has not enough native asset, trigger swap from dex
+				let treasury_balance = T::Currency::free_balance(&treasury_account);
+				if treasury_balance < SwapBalanceThreshold::<T>::get(supply_currency_id) {
+					let trading_path = Self::get_trading_path_by_currency(&treasury_account, supply_currency_id);
+					if let Some(trading_path) = trading_path {
+						// using all supply asset to swap out native asset, no need to consider ED limitation here.
+						let treasury_supply_balance =
+							T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+						if let Ok(swap_native_balance) =
+							T::DEX::swap_with_exact_supply(&treasury_account, &trading_path, treasury_supply_balance, 0)
 						{
-							// successfully swap, break iteration
-							break;
+							// calculate and update new rate of supply asset relative to native asset
+							let new_native_balance =
+								rate.saturating_mul_int(swap_native_balance.saturating_add(treasury_balance));
+							rate = Ratio::saturating_from_rational(new_native_balance, PoolSize::<T>::get());
```

Comment:
I think it is ok to make this transaction to use the old rate. makes frontend calculation easier.

This formula means we can't change pool size otherwise the rate will be wrong.